File: sdk/hcs-u7-python/src/hcs_u7/parser.py
from __future__ import annotations

import re
from typing import Dict

from .types import HCSProfile, ModalDistribution, CognitionScores, InteractionSettings


class HCSCodeParseError(ValueError):
    """Raised when an HCS-U7 code is invalid or cannot be parsed."""


def _parse_segments(code: str) -> Dict[str, str]:
    parts = code.split("|")
    if len(parts) < 2 or not parts[0].startswith("HCS-U7"):
        raise HCSCodeParseError("Invalid HCS-U7 header")

    segments: Dict[str, str] = {}
    for seg in parts[1:]:
        if ":" in seg:
            key, value = seg.split(":", 1)
            segments[key] = value
    return segments


def _parse_modal(value: str | None) -> ModalDistribution:
    modal = ModalDistribution()
    if not value:
        return modal

    for key, num in re.findall(r"([cfm])(\d+)", value):
        val = int(num)
        if key == "c":
            modal.c = val
        elif key == "f":
            modal.f = val
        elif key == "m":
            modal.m = val
    return modal


def _parse_cognition(value: str | None) -> CognitionScores:
    scores = {"form": 0, "logic": 0, "visual": 0, "synthesis": 0, "creative": 0}
    if not value:
        return CognitionScores(**scores)

    mapping = {"F": "form", "C": "logic", "V": "visual", "S": "synthesis", "Cr": "creative"}
    for code, num in re.findall(r"([A-Z][a-z]?)(\d+)", value):
        key = mapping.get(code)
        if key:
            scores[key] = int(num)

    return CognitionScores(**scores)


def _parse_interaction(value: str | None) -> InteractionSettings:
    settings = InteractionSettings()
    if not value:
        return settings

    for part in value.split(","):
        if "=" not in part:
            continue
        raw_key, raw_val = part.split("=", 1)
        key = raw_key.strip()
        val = raw_val.strip()
        if key in {"PB", "SM", "TN"}:
            setattr(settings, key, val)
    return settings


def parse_hcs_code(code: str) -> HCSProfile:
    """Parse an HCS-U7 code into a structured HCSProfile instance."""

    trimmed = code.strip()
    if not trimmed:
        raise HCSCodeParseError("Empty HCS-U7 code")

    segments = _parse_segments(trimmed)

    version = segments.get("V", "7.0")
    algorithm = segments.get("ALG", "QS")
    element = segments.get("E", "E")
    modal = _parse_modal(segments.get("MOD"))
    cognition = _parse_cognition(segments.get("COG"))
    interaction = _parse_interaction(segments.get("INT"))
    qsig = segments.get("QSIG")
    b3 = segments.get("B3")

    return HCSProfile(
        raw_code=trimmed,
        version=version,
        algorithm=algorithm,
        element=element,
        modal=modal,
        cognition=cognition,
        interaction=interaction,
        qsig=qsig,
        b3=b3,
    )

File: sdk/hcs-u7-js/src/parser.ts
import type { HCSProfile, ModalDistribution, CognitionScores, InteractionSettings } from "./types";

export class HCSCodeParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "HCSCodeParseError";
  }
}

function parseSegments(code: string): Record<string, string> {
  const parts = code.split("|");
  if (parts.length < 2 || !parts[0].startsWith("HCS-U7")) {
    throw new HCSCodeParseError("Invalid HCS-U7 header");
  }

  const segments: Record<string, string> = {};
  for (const part of parts.slice(1)) {
    const idx = part.indexOf(":");
    if (idx === -1) continue;
    const key = part.slice(0, idx);
    const value = part.slice(idx + 1);
    segments[key] = value;
  }

  return segments;
}

function parseModal(value: string | undefined): ModalDistribution {
  const modal: ModalDistribution = { c: 0, f: 0, m: 0 };
  if (!value) return modal;

  const regex = /([cfm])(\d+)/g;
  let match: RegExpExecArray | null;

  while ((match = regex.exec(value)) !== null) {
    const [, key, num] = match;
    const n = parseInt(num, 10);
    if (key === "c") modal.c = n;
    if (key === "f") modal.f = n;
    if (key === "m") modal.m = n;
  }

  return modal;
}

function parseCognition(value: string | undefined): CognitionScores {
  const scores: CognitionScores = {
    form: 0,
    logic: 0,
    visual: 0,
    synthesis: 0,
    creative: 0,
  };

  if (!value) return scores;

  const regex = /([A-Z][a-z]?)(\d+)/g;
  const map: Record<string, keyof CognitionScores> = {
    F: "form",
    C: "logic",
    V: "visual",
    S: "synthesis",
    Cr: "creative",
  };

  let match: RegExpExecArray | null;
  while ((match = regex.exec(value)) !== null) {
    const [, code, num] = match;
    const n = parseInt(num, 10);
    const key = map[code];
    if (key) {
      scores[key] = n;
    }
  }

  return scores;
}

function parseInteraction(value: string | undefined): InteractionSettings {
  const interaction: InteractionSettings = {};
  if (!value) return interaction;

  for (const part of value.split(",")) {
    const [rawKey, rawVal] = part.split("=", 2);
    if (!rawKey || !rawVal) continue;
    interaction[rawKey.trim()] = rawVal.trim();
  }

  return interaction;
}

export function parseHCSCode(raw: string): HCSProfile {
  const code = raw.trim();
  if (!code) {
    throw new HCSCodeParseError("Empty HCS-U7 code");
  }

  const segments = parseSegments(code);

  const version = segments["V"] ?? "7.0";
  const algorithm = segments["ALG"] ?? "QS";
  const element = segments["E"] ?? "E";
  const modal = parseModal(segments["MOD"]);
  const cognition = parseCognition(segments["COG"]);
  const interaction = parseInteraction(segments["INT"]);
  const qsig = segments["QSIG"];
  const b3 = segments["B3"];

  return {
    rawCode: code,
    version,
    algorithm,
    element,
    modal,
    cognition,
    interaction,
    qsig,
    b3,
  };
}
